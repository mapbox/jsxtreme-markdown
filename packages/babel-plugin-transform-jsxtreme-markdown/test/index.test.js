'use strict';

const babel = require('@babel/core');
const babelPresetEnv = require('@babel/preset-env');
const plugin = require('../index.js');

const transform = (code, options) => {
  return babel.transform(code, {
    presets: [babelPresetEnv],
    plugins: [[plugin, options]],
    configFile: false
  }).code;
};

test('basic usage', () => {
  const code = `
    const md = require('@mapbox/babel-plugin-transform-jsxtreme-markdown/md');
    const foo = md\`
      # Title
      This is **bold.**
      Here is a [link](/some/url).
    \`;
  `;
  expect(transform(code)).toMatchSnapshot();
});

test('import instead of require', () => {
  const code = `
    import md from '@mapbox/babel-plugin-transform-jsxtreme-markdown/md';
    const foo = md\`
      # Title
      This is **bold.**
      Here is a [link](/some/url).
    \`;
  `;
  expect(transform(code)).toMatchSnapshot();
});

test('alternate package name', () => {
  const code = `
    import md from 'markdownify';
    const foo = md\`
      # Title
      This is **bold.**
      Here is a [link](/some/url).
    \`;
  `;
  expect(transform(code, { packageName: 'markdownify' })).toMatchSnapshot();
});

test('require in nested scope', () => {
  const code = `
    function x() {
      if (true) {
        const md = require('@mapbox/babel-plugin-transform-jsxtreme-markdown/md');
        const foo = md\`
          # Title
          This is **bold.**
          Here is a [link](/some/url).
        \`;
      }
    }
  `;
  expect(transform(code)).toMatchSnapshot();
});

test('alternate variable name, with broken-up nested JSX', () => {
  const code = `
    const markit = require('@mapbox/babel-plugin-transform-jsxtreme-markdown/md');
    const text = markit\`
    This is a paragraph {{ <span className="foo"> }} with a **markdown** span inside {{ </span> }}
    {{ <div style={{ margin: 70 }}> }}
      And here is a *paragraph* inside a div.
      [Link](/some/url)
    {{ </div> }}\`
  `;
  expect(transform(code)).toMatchSnapshot();
});

test('toJsx options', () => {
  const code = `
    import md from '@mapbox/babel-plugin-transform-jsxtreme-markdown/md';
    const foo = md\`
      This is a paragraph {{ <span className="foo">with a span inside</span> }}
      {# <div style={{ margin: 70 }}> #}
        And **here*** is a div with {# Markdown #} inside.
      {# </div> #}
    \`;
  `;
  expect(
    transform(code, {
      delimiters: ['{#', '#}']
    })
  ).toMatchSnapshot();
});

test('fails with placeholders in template literals', () => {
  const code = `
    const md = require('@mapbox/babel-plugin-transform-jsxtreme-markdown/md');
    const foo = md\`
      # Title
      This is **bold.** $\{number}
      Here is a [link](/some/url).
    \`;
  `;
  expect(() => transform(code)).toThrow('Placeholders are not supported');
});

test('does nothing when module is not in scope', () => {
  const code = `
    function x() {
      const md = require('@mapbox/babel-plugin-transform-jsxtreme-markdown/md');
    }
    const foo = md\`
      # Title
      This is **bold.**
      Here is a [link](/some/url).
    \`;
  `;
  expect(transform(code)).toMatchSnapshot();
});

test('does not add React to scope when it is already required', () => {
  const code = `
    const React = require('react');
    const md = require('@mapbox/babel-plugin-transform-jsxtreme-markdown/md');
    const foo = md\`
      # Title
      This is **bold.**
      Here is a [link](/some/url).
    \`;
  `;
  expect(transform(code)).toMatchSnapshot();
});

test('does not add React to scope when it is already imported', () => {
  const code = `
    import React from 'react';
    import md from '@mapbox/babel-plugin-transform-jsxtreme-markdown/md';
    const foo = md\`
      # Title
      This is **bold.**
      Here is a [link](/some/url).
    \`;
  `;
  expect(transform(code)).toMatchSnapshot();
});
