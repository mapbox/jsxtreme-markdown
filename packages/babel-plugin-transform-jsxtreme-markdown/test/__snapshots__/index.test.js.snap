// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`alternate package name 1`] = `
"\\"use strict\\";

var foo = <div><h1>Title</h1>
  <p>This is <strong>bold.</strong>
Here is a <a href=\\"/some/url\\">link</a>.</p></div>;"
`;

exports[`alternate variable name, with broken-up nested JSX 1`] = `
"\\"use strict\\";

var text = <div><p>This is a paragraph <span className=\\"foo\\"> with a <strong>markdown</strong> span inside </span></p>
  <div style={{
    margin: 70
  }}>
    <p>      And here is a <em>paragraph</em> inside a div.
      <a href=\\"/some/url\\">Link</a></p>
  </div></div>;"
`;

exports[`basic usage 1`] = `
"\\"use strict\\";

var foo = <div><h1>Title</h1>
  <p>This is <strong>bold.</strong>
Here is a <a href=\\"/some/url\\">link</a>.</p></div>;"
`;

exports[`does not add React to scope when it is already imported 1`] = `
"\\"use strict\\";

var _react = _interopRequireDefault(require(\\"react\\"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { \\"default\\": obj }; }

var foo = <div><h1>Title</h1>
  <p>This is <strong>bold.</strong>
Here is a <a href=\\"/some/url\\">link</a>.</p></div>;"
`;

exports[`does not add React to scope when it is already required 1`] = `
"\\"use strict\\";

var React = require('react');

var foo = <div><h1>Title</h1>
  <p>This is <strong>bold.</strong>
Here is a <a href=\\"/some/url\\">link</a>.</p></div>;"
`;

exports[`does nothing when module is not in scope 1`] = `
"\\"use strict\\";

function _templateObject() {
  var data = _taggedTemplateLiteral([\\"\\\\n      # Title\\\\n      This is **bold.**\\\\n      Here is a [link](/some/url).\\\\n    \\"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

function x() {}

var foo = md(_templateObject());"
`;

exports[`import instead of require 1`] = `
"\\"use strict\\";

var foo = <div><h1>Title</h1>
  <p>This is <strong>bold.</strong>
Here is a <a href=\\"/some/url\\">link</a>.</p></div>;"
`;

exports[`require in nested scope 1`] = `
"\\"use strict\\";

function x() {
  if (true) {
    var foo = <div><h1>Title</h1>
      <p>This is <strong>bold.</strong>
Here is a <a href=\\"/some/url\\">link</a>.</p></div>;
  }
}"
`;

exports[`toJsx options 1`] = `
"\\"use strict\\";

var foo = <div><p>This is a paragraph {\\"{\\"}{\\"{\\"} <span className=\\"foo\\">with a span inside</span> {\\"}\\"}{\\"}\\"}</p>
  <div style={{
    margin: 70
  }}>
    <p>        And <strong>here*</strong> is a div with {Markdown} inside.</p>
  </div></div>;"
`;
